Common > Validation: 유효성 응답 모델에 "datetime:{variant}" 타입 프로퍼티 모델들 추가

merge-simpson opened the pull request:
<!--
<br /><br /><br /><br /><br />

<table align="center" border="30">
<tr>
  <td>🏗️ 선행 작업 병합 후, 리베이스할 예정입니다.</td>
</tr>
</table>

<br /><br /><br /><br /><br />
그니까 아직 리뷰하지 말라니깐용.
<br /><br /><br /><br /><br />
<br /><br /><br /><br /><br />
-->

## Issues

- Resolves #332 
- Resolves #333 

<br />

## Description

- Datetime 데이터의 유효성 모델을 추가합니다.
- Rel: #323 

<br />

## Review Points

- 노션 스펙을 참고하실 수 있습니다. (로그인 필요)  
  https://www.notion.so/nettee/Auth-User-API-25001973b23a800c9e9fe2392aa7fa97#25001973b23a802aad45ea139d9f1743
- 관련 파일을 참고하실 수 있습니다.  
  - common/src/main/java/nettee/common/validation/model/interfaces/**BaseValidationProperty**.java  
    ```java
    public interface BaseValidationProperty {
        String type();
        Boolean required();
        Map<String, String> messages();
        
        // ... (중략)
    }
    ```  
  - common/src/main/java/nettee/common/validation/model/**ValidationResponseModel**.java

<br />

## How Has This Been Tested?

- 테스트 생략

<br />

## Additional Notes

### 다이어그램

```mermaid
classDiagram
direction TB

%% ================= Base & Common =================
class BaseValidationProperty {
  <<interface>>
  +type() String
  +required() Boolean
  +messages() Map~String,String~
}

class DateTimeBaseValidationProperty {
  <<interface>>
}

BaseValidationProperty <|-- DateTimeBaseValidationProperty

%% ================= Interfaces for datetime =================
class SinceUntilDateTimeValidationProperty {
  <<interface>>
  +since() Instant
  +until() Instant
  +inclusive() List~Boolean~
}
DateTimeBaseValidationProperty <|-- SinceUntilDateTimeValidationProperty

class SteppedDateTimeValidationProperty {
  <<interface>>
  +step() String
  +stepEpoch() String
}
DateTimeBaseValidationProperty <|-- SteppedDateTimeValidationProperty

class DurationBoundedValidationProperty {
  <<interface>>
  +minDuration() DurationValidationSpec
  +maxDuration() DurationValidationSpec
  +inclusive() List~Boolean~
}
DateTimeBaseValidationProperty <|-- DurationBoundedValidationProperty

%% ================= Records (implementations) =================
class AbsoluteDateTimeValidationProperty {
  <<record>>
  +type() String
  +required() Boolean
  +since() Instant
  +until() Instant
  +inclusive() List~Boolean~
  +step() String
  +stepEpoch() String
  +messages() Map~String,String~
}
SinceUntilDateTimeValidationProperty <|.. AbsoluteDateTimeValidationProperty
SteppedDateTimeValidationProperty <|.. AbsoluteDateTimeValidationProperty

class RelativeDateTimeValidationProperty {
  <<record>>
  +type() String
  +required() Boolean
  +minDuration() DurationValidationSpec
  +maxDuration() DurationValidationSpec
  +inclusive() List~Boolean~
  +step() String
  +stepEpoch() String
  +messages() Map~String,String~
}
DurationBoundedValidationProperty <|.. RelativeDateTimeValidationProperty
SteppedDateTimeValidationProperty <|.. RelativeDateTimeValidationProperty
DurationValidationSpec <-- RelativeDateTimeValidationProperty : uses

%% ================= Duration spec =================
class DurationValidationSpec {
  <<record>>
  +ref() String
  +offset() String
}
DurationValidationSpec <-- DurationBoundedValidationProperty : returns
```

merge-simpson committed:
feat(common): add duration validation spec with datetime `ref` and iso-8601 `offset`

Introduce DurationValidationSpec record supporting reserved ref types and resource field references.
Validate `ref` against supported enums or resource-prefixed names and enforce `offset` to follow ISO-8601 duration format.

merge-simpson committed:
feat(common): add field name constants `SINCE`, `UNTIL` for absolute datetime validation

merge-simpson committed:
feat(common): introduce base interface for datetime validation models

Add `DateTimeBaseValidationProperty` extending `BaseValidationProperty` to standardize contracts for datetime-related validations. This prepares for absolute/relative datetime validation implementations and does not introduce runtime behavior changes.

merge-simpson committed:
feat(common): add duration bounded validation property interface for datetime validation model

Introduce `DurationBoundedValidationProperty` extending `DateTimeBaseValidationProperty` with `minDuration`, `maxDuration`, and `inclusive` methods.
This provides a contract for validations requiring bounded duration ranges in datetime properties.

merge-simpson committed:
feat(common): add since-until datetime validation property interface

Introduce `SinceUntilDateTimeValidationProperty` extending `DateTimeBaseValidationProperty` with `since`, `until`, and `inclusive` methods.
This defines a contract for validations requiring fixed datetime boundaries.

merge-simpson committed:
feat(common): add stepped datetime validation property interface

Introduce `SteppedDateTimeValidationProperty` extending `DateTimeBaseValidationProperty` with `step` and `stepEpoch` methods.
This defines a contract for validations that enforce datetime stepping relative to a reference epoch.

merge-simpson committed:
feat(common): implement absolute datetime validation property record

Add `AbsoluteDateTimeValidationProperty` implementing `SinceUntilDateTimeValidationProperty` and `SteppedDateTimeValidationProperty`.
Includes validation for type consistency, `since`/`until` bounds, `inclusive` size, and `step`-`stepEpoch` dependency.
Also initializes default values and auto-populates validation messages for `required`, `since`, and `until` fields.

merge-simpson committed:
feat(common): implement relative datetime validation property record

Add `RelativeDateTimeValidationProperty` implementing `DurationBoundedValidationProperty` and `SteppedDateTimeValidationProperty`.
Includes validation for `type` consistency, duration `ref` matching, inclusive size, and `step`-`stepEpoch` dependency.
Also normalizes blank values, initializes defaults for `type`, `required`, and `inclusive`, and auto-populates validation `messages` for `required`, `minDuration`, and `maxDuration`.

wowddok99 commented on common/src/main/java/nettee/common/validation/model/AbsoluteDateTimeValidationProperty.java:
```diff
@@ -0,0 +1,68 @@
+        if (required && !messages.containsKey(REQUIRED)) {
+            messages.put(REQUIRED, "필수 입력 항목입니다.");
+        }
+        if (since != null && messages.containsKey(SINCE)) {
```
```java
if (since != null && !messages.containsKey(SINCE)) {
            messages.put(SINCE, since + " 이후만 선택할 수 있습니다.");
```
- messages.containsKey(SINCE)앞에 느낌표가 누락된듯 합니다!

wowddok99 commented on common/src/main/java/nettee/common/validation/model/AbsoluteDateTimeValidationProperty.java:
```diff
@@ -0,0 +1,68 @@
+        if (since != null && messages.containsKey(SINCE)) {
+            messages.put(SINCE, since + " 이후만 선택할 수 있습니다.");
+        }
+        if (until != null && messages.containsKey(UNTIL)) {
```
위와 동일하게 느낌표가 누락된듯 합니다 : )

wowddok99 commented on common/src/main/java/nettee/common/validation/model/interfaces/DurationValidationSpec.java:
```diff
@@ -0,0 +1,64 @@
+                                .map(Enum::name)
+                                .<String>reduce((acc, cur) -> acc + ", " + cur) +
+                        ", or " + RESOURCE_FIELD_REF_PREFIX + "<fieldName> instead." ;
+        assert offset.matches("^P(?!$)(\\d+Y)?(\\d+M)?(\\d+D)?(T(\\d+H)?(\\d+M)?(\\d+S)?)?$")
```
```java
assert offset.matches("^P(?!$)(\\d+Y)?(\\d+M)?(\\d+D)?(T(\\d+H)?(\\d+M)?(\\d+S)?)?$")
: "step must be a valid ISO-8601 duration (e.g. PT5M, P1D, P2W, P3M)"
```
현재 실제 정규식에는 W(주 단위)가 포함되어 있지 않은데, 에러메시지에서는 P2W를 예시로 제시하고 있어서 코드와 메시지가 불일치하는 상황으로 보입니다. 한번 확인 정도만 해주시면 감사하겠습니다 : )

merge-simpson replied:
어떻게 발견하셨는지 대단합니다. 감사합니다.
혼란이 없도록 프론트엔드에 제공하는 정규식을 갖고 와서 넣어야겠습니다.

wowddok99 reviewed (approved):
올려주신 코드 확인하였습니다.
간단한 코멘트는 남겨두었고 확인 이후에 변경할 부분이 있다면 변경해주시면 될듯합니다!
우선 Approve 해두겠습니다. 작업 고생하셨어요😁

merge-simpson committed:
fix(common): fix inverted message key check for `since`/`until` in datetime validation properties

correct condition from `messages.containsKey(UNTIL)` to `!messages.containsKey(UNTIL)` (and same for `SINCE`) so default messages apply only when caller does not provide them; updated in `AbsoluteDateTimeValidationProperty` and `RelativeDateTimeValidationProperty`.

merge-simpson committed:
fix(common): fix asserted regexp for `ref` in `DurationValidationSpec`

merge-simpson commented:
발견해 주신 모든 실수를 수정하여 병합합니다.

@wowddok99 님, 감사합니다.
